src/mds/QuiesceDb.h

```diff
@@ -0,0 +1,338 @@
+#include <string>
+#include <map>
+#include <unordered_map>
+#include <optional>
+#include <vector>
+
+#include "mds/mdstypes.h"
+#include "common/ceph_time.h"
+
+
+// NB! The order of the states in the enum is important!
+// There are places in the code that aggregate multiple states
+// via min or max, depending on the task.
+// The order of states represents the natural lifecycle
+// of a set and its members, this is specifically important
+// for the active states.
+enum QuiesceState {
+  QS__INVALID,
+
+  // these states are considered "active"
+  QS_QUIESCING, QS__ACTIVE = QS_QUIESCING,
+  QS_QUIESCED,
+  QS_RELEASING,
+
+  // the below states are all terminal, or "inactive"
+  QS_RELEASED, QS__TERMINAL = QS_RELEASED,
+  QS_EXPIRED,
+
+  // the below states aren't allowed for roots, only for sets
+  QS_CANCELED, QS__SET_ONLY = QS_CANCELED,
+  QS_TIMEDOUT,
+
+  QS__MAX,
+};
+
+template <class CharT, class Traits>
+static std::basic_ostream<CharT, Traits>&
+operator<<(std::basic_ostream<CharT, Traits>& os, const QuiesceState& qs)
+{
+  switch (qs) {
+  case QS__INVALID:
+    return os << "QS__INVALID (" << (int)qs << ")";
+  case QS_QUIESCING:
+    return os << "QS_QUIESCING (" << (int)qs << ")";
+  case QS_QUIESCED:
+    return os << "QS_QUIESCED (" << (int)qs << ")";
+  case QS_RELEASING:
+    return os << "QS_RELEASING (" << (int)qs << ")";
+  case QS_RELEASED:
+    return os << "QS_RELEASED (" << (int)qs << ")";
+  case QS_CANCELED:
+    return os << "QS_CANCELED (" << (int)qs << ")";
+  case QS_TIMEDOUT:
+    return os << "QS_TIMEDOUT (" << (int)qs << ")";
+  case QS_EXPIRED:
+    return os << "QS_EXPIRED (" << (int)qs << ")";
+  default:
+    return os << "!Unknown quiesce state! (" << (int)qs << ")";
+  }
+};
+
+// Since MDS clock is not syncrhonized, and the quiesce db has to be replicated,
+// we measure all events in the quiesce database relative to the database age.
+// The age of the database is just large enough to have earliest events carry
+// a non-negative age stamp.
+// This is sufficient because we only care to honor the timeouts that are relative
+// to the other recorded database events.
+// This approach also relieves us from storing or transfering absolute time stamps:
+// every client can deduce the lower boundary of event's absolute time given the 
+// message roundrip timing - if they bother enough. Otherwise, they can just subtract
+// the received database age from now() and get their own absolute time reference.
+
+using QuiesceClock = ceph::coarse_real_clock;
+using QuiesceTimePoint = QuiesceClock::time_point;
+using QuiesceTimeInterval = QuiesceClock::duration;
+using QuiesceSetId = std::string;
+using QuiesceRoot = std::string; // TODO: file path?
+using QuiesceDbVersion = uint32_t;
+
+struct RecordedQuiesceState {
+  QuiesceState state;
+  QuiesceTimeInterval at_age;
+
+  operator QuiesceState() {
+    return state;
+  }
+
+  bool update(const QuiesceState state, const QuiesceTimeInterval at_age) {
```

Comment:
I agree, a little odd. I don't see any reason to make it a reference, as you pointed out both are primitive types. I added const to prevent accidental mutation of the parameters instead of the fields, given that they have the same name.

Reply:
I went for the const reference. Let the compiler do it job.